hashintel/hash
# H-466: Update how we detect link types to account for inheritance

## 🌟 What is the purpose of this PR?

In various places in the app we want to know if an entity type is a link type, i.e. represents a class of entities that define relationships between other entities. Our marker for an entity type being a link type is that it inherits from the special Block Protocol `Link` type.

Prior to inheritance, we checked whether or not a type was a link type by looking in its `allOf` for the presence of `Link`. Since an entity type may now _inherit_ its link status by virtue of one of its ancestors inheriting from `Link`, we need to update our checks accordingly. This PR does that.




## 🔍 What does this change?

- Updates `isEntityTypeLinkEntityType` in the node API
- Adds `isLinkTypeLookup` to the entity type context in the frontend (so that we only need to calculate it when we first fetch all the entity types), replace existing checks with it
- Adds link types into the possible parents in the type editor (previously we weren't passing it which meant you couldn't set or see parents which inherited from `Link`)
- Updates the code in the type editor where we split types into entity or link


## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues

- It is possible to add a non-link type as a parent of a link type. We want to restrict this at first, and I will do so in a follow-up PR which adds other validation when choosing a parent (e.g. to prevent cycles and conflicting properties in the chain)


## 🐾 Next steps

- Add validation when users choose a parent


## 🛡 What tests cover this?



- None

## ❓ How to test this?



1.  Create a link type (a type which inherits from link), and then create another type which extends that one
2. Check that the second type appears as a link (should have a link icon in the header, appear as an option when selecting a link to add to other types)
